### PR TITLE
Improve BBolt locking

### PR DESCRIPTION
**Pull request by reinkrul**
- Improved TX option invocation design
- Fixed TX lock unlocking for Redis; previously happened after AfterCommit/OnRollback invocations, now happens before (found when enabling BBolt lock expiry test)
- Introduced configurable lock acquisition time-out (default 3s)
- Introduced lock cancellation for BBolt

### CODE REVIEW

File: store.go [hunk 1]
```diff
@@ -126,31 +143,13 @@ type Store interface {
...
-		if ar, ok := opt.(*OnRollbackOpt); ok {
-			ar.Func()
-		}
-	}
+type WriteLockOption struct {
 }
 
-func (opts TxOptions) InvokeAfterCommit() {
+// Enabled returns whether the WithWriteLock option was specified.
+func (w WriteLockOption) Enabled(opts []TxOption) bool {
```
Comment (woutslakhorst) [inline store.go:151]: Why is this a method on `WriteLockOption`?

File: util/mutex.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
+package util
```
Comment (woutslakhorst) [inline util/mutex.go:1]: copyright

File: util/mutex.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	go func() {
+		fnLock()
+		m.Lock()
+		if expired.Load() != nil {
+			// Context expired, unlock immediately.
+			fnUnlock()
+		} else {
+			locked <- true
+		}
+		m.Unlock()
```
Comment (woutslakhorst) [inline util/mutex.go:59]: usually it's a `defer` right after the lock

File: util/mutex_test.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
+package util
+
+import (
+	"context"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func Test_ContextRWLocker(t *testing.T) {
```
Comment (woutslakhorst) [inline util/mutex_test.go:9]: Expected a test which with a context deadline exceeded.

File: kvtests/common_tests.go [hunk 5]
```diff
@@ -598,7 +597,8 @@ func TestWriteTransactions(t *testing.T, storeProvider StoreProvider) {
 func TestTransactionWriteLock(t *testing.T, storeProvider StoreProvider) {
 	ctx := context.Background()
 	supportsLockExpiry := func(store stoabs.KVStore) bool {
-		return fmt.Sprintf("%T", store) != "*bbolt.store"
+		return true
+		//return fmt.Sprintf("%T", store) != "*bbolt.store"
```
Comment (woutslakhorst) [inline kvtests/common_tests.go:601]: leftover comment

**Review by woutslakhorst — CHANGES_REQUESTED** [review]

File: util/mutex.go (hunk 3 above)
Comment (reinkrul) [reply util/mutex.go:59]: I guess the few extra nanoseconds `defer` adds won't really hurt, given the lock is used for securing expensive database access.

File: util/mutex.go (hunk 2 above)
Comment (reinkrul) [reply util/mutex.go:1]: Ah yes, the thing I always manage to forget about.

File: store.go (hunk 1 above)
Comment (reinkrul) [reply store.go:151]: I figured it is a better pattern than having a `TxOptions` (alias for a slice) that has all these option-specific functions. Now it's function to read it from an options list is bound to the option, at least.

File: store.go [hunk 6]
```diff
@@ -31,6 +31,8 @@ var ErrStoreIsClosed = errors.New("database not open")
 
 const DefaultTransactionTimeout = 30 * time.Second
 
+const defaultLockAcquisitionTimeout = 3 * time.Second
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply store.go:34]: It's kinda arbitrary, but given this library is intended to be used outside of nuts-node as well, I figured 1 second might be a bit short. We can still set it to 1s when initializing the store in nuts-node using `WithLockAcquireTimeout`?

**Review by woutslakhorst — APPROVED** [review]

**Pull request merged**